File: spaceCadets-whatABook.py
#-----------------------------------------------------
# Title: spaceCadets-whatABook.py
# Author: Danielle Taplin, Nolan Berryhill
# Date: 01 October 2023
# Description: Python for WhatABook
#-----------------------------------------------------

import pymongo

# Connect to MongoDB
client = pymongo.MongoClient("mongodb+srv://whatabook_user:s3cret@cluster0.wmphxtw.mongodb.net/")

# Access the "books" collection
db = client.WhatABook
books_collection = db.books

# Display a list of Books
def display_books():
    for book in books_collection.find():
        print(f"Title: {book['title']}, Genre: {book['genre']}, Author: {book['author']}")

def display_books_by_genre():
    # Specify the genre
    genre = input("Enter genre: ")

    # Display a list of books by genre
    for book in books_collection.find({"genre": genre}):
        print(f"Title: {book['title']}, Author: {book['author']}")

def display_books_by_author():
    # Specify the author
    author = input("Enter author: ")

    # Display a list books by author
    for book in books_collection.find({"author": author}):
        print(f"Title: {book['title']}, Genre: {book['genre']}")

def display_book_by_id():
    # Specify the bookId
    book_id = input("Enter bookId: ")

    # Display a book by bookId
    book = books_collection.find_one({"bookId": book_id})
    if book:
        print(f"Title: {book['title']}, Genre: {book['genre']}, Author: {book['author']}")
    else:
        print(f"Book with bookId {book_id} not found.")

#Menu-driven interface
while True:
    print('\nMenu:')
    print("1. Display a list of books")
    print("2. Display a list of books by genre")
    print("3. Display a list of books by author")
    print("4. Display a book by bookId")
    print("5. Exit")

    choice = input("Enter your choice: ")

    if choice == "1":
        display_books()
    elif choice == "2":
        display_books_by_genre()
    elif choice == "3":
        display_books_by_author()
    elif choice == "4":
        display_book_by_id()
    elif choice == "5":
        break
    else:
        print("Invalid choice. Please select a valid option.")

# Close the MongoDB connection
client.close()

File: spaceCadets-whatABook.js
/**
	Title: spaceCadets-whatABooks.js
    Author: Danielle Taplin, Nolan Berryhill
    Date: 01 October 2023
    Description: Script for WhatABook
 */

// Import the MongoDB driver
const { MongoClient } = require('mongodb');

// Connection URL
const url = 'mongodb+srv://whatabook_admin:s3cret@cluster0.wmphxtw.mongodb.net/';
const dbName = 'WhatABook';

// Create a new MongoClient instance
const client = new MongoClient(url, { useNewUrlParser: true, useUnifiedTopology: true });

// Function to initialize the database
async function initDatabase() {
    try {
        // Connect to the MongoDB server
        await client.connect();
        const db = client.db(dbName);

        // Check if collections exist before dropping
        const collections = await db.listCollections().toArray();
        const existingCollections = collections.map((collection) => collection.name);

        if (existingCollections.includes('books')) {
            await db.dropCollection('books').drop();
        }

        if (existingCollections.includes('customers')) {
            await db.dropCollection('customers').drop();
        }

        if (existingCollections.includes('wishlistitems')) {
            await db.dropCollection('wishlistitems').drop();
        }

        // Drop existing collections
        await db.dropCollection('books');
        await db.dropCollection('customers');
        await db.dropCollection('wishlistitems');

        // Create collections
        const booksCollection = db.collection('books');
        const customersCollection = db.collection('customers');
        const wishlistItemsCollection = db.collection('wishlistitems');

        // Create the "User" collection
        const userCollection = db.collection('User');

        // Insert data into collections
        await booksCollection.insertMany(booksData);
        await customersCollection.insertMany(customersData);
        await wishlistItemsCollection.insertMany(wishlistItemsData);

        console.log('Database initialized.');
    } catch (err) {
        console.error('Error initializing database:', err);
    } finally {
        client.close();
    }
};

// Book list
const booksData = [
    { bookId: 's1001', title: 'The Lincoln Lawyer', genre: 'Mystery', author: 'Micheal Connelly' },
    { bookId: 's1002', title: 'Foreign Deceit', genre: 'Mystery', author: 'Jeff Carson' },
    { bookId: 's1003', title: 'Red Knife', genre: 'Mystery', author: 'William Krueger' },
    { bookId: 's1004', title: 'The Wager: A Tale of Shipwreck', genre: 'Mystery', author: 'David Grann' },
    { bookId: 's1005', title: 'Way of the Wolf', genre: 'Fantasy', author: 'E. Knight' },
    { bookId: 's1006', title: 'Death Masks', genre: 'Fantasy', author: 'Jim Butcher' },
    { bookId: 's1007', title: 'Apocalyptic Space', genre: 'Fantasy', author: 'Stephen Lee' },
    { bookId: 's1008', title: 'Scion of Conquered Earth', genre: 'Fantasy', author: 'Micheal Allen' },
    { bookId: 's1009', title: 'Twenty Years Later', genre: 'Thriller', author: 'Charlie Donlea' },
    { bookId: 's1010', title: 'Legacy of Lies', genre: 'Thriller', author: 'Robert Bailey' },
    { bookId: 's1011', title: 'One by One', genre: 'Thriller', author: 'Freida Mcfadden' },
    { bookId: 's1012', title: 'The family across the street', genre: 'Thriller', author: 'Nicole Trope' }
];

// Customer information
const customersData = [
    { customerId: 'c1007', firstName: 'Olivia', lastName: 'Thompson' },
    { customerId: 'c1008', firstName: 'Mark', lastName: 'Anderson' },
    { customerId: 'c1009', firstName: 'Bob', lastName: 'Builder' },
];

// Wish list for customer
const wishlistItemsData = [
    { customerId: 'c1007', bookId: ['s1009', 's1010', 's1011', 's1012'] },
    { customerId: 'c1008', bookId: ['s1005', 's1006', 's1007', 's1008'] },
    { customerId: 'c1009', bookId: ['s1001', 's1002', 's1003', 's1004'] },
];

// Call the initDatabase function with the client
initDatabase();